File: spring-data-requery/src/test/java/org/springframework/data/requery/domain/sample/AbstractVersionedUser.java
package org.springframework.data.requery.domain.sample;

import io.requery.*;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.requery.domain.AbstractPersistable;
import org.springframework.data.requery.domain.ToStringBuilder;

import java.util.Date;
import java.util.Objects;

/**
 * AbstractVersionedUser
 *
 * @author debop@coupang.com
 * @since 18. 6. 25
 */
@Entity
public abstract class AbstractVersionedUser extends AbstractPersistable<Long> {

    @Key
    @Generated
    protected Long id;

    @Version
    protected Long version;


    protected String name;

    protected String email;

    protected Date birthday;

    public AbstractVersionedUser() {}

    public AbstractVersionedUser(String name, String email, Date birthday) {
        this.name = name;
        this.email = email;
        this.birthday = birthday;
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, email, birthday);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name)
            .add("email", email)
            .add("birthday", birthday);
    }

    private static final long serialVersionUID = -7900602652543174851L;
}

File: spring-data-requery/src/test/kotlin/org/springframework/data/requery/domain/sample/AbstractVersionedUser.kt
package org.springframework.data.requery.domain.sample

import io.requery.*
import org.springframework.data.requery.domain.AbstractPersistable
import org.springframework.data.requery.domain.ToStringBuilder
import java.util.*

/**
 * org.springframework.data.requery.domain.sample.AbstractVersionedUser
 *
 * @author debop
 */
@Entity
abstract class AbstractVersionedUser: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    abstract override val id: Long

    @get:Column(name = "user_name")
    abstract var name: String

    @get:Column(name = "user_email")
    abstract var email: String

    @get:Column(name = "user_birthday")
    abstract var birthday: Date

    //NOTE: @Version 을 지정하기 위해서는 @JvmField를 지정하고, final로 지정해야 합니다.
    @Version
    @JvmField
    final var version: Long = 0

    override fun hashCode(): Int = Objects.hash(name, email, birthday)

    @io.requery.Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("name", name)
            .add("email", email)
            .add("birthday", birthday)
    }
}